flow/project.py

```diff
@@ -687,6 +702,89 @@ def label_func(func):
 
         return label_func
 
+    def detect_operation_graph(self):
+        """Determine the directed acyclic graph defined by operation pre- and
+        post- conditions.
+
+        In general, executing a given operation registered with a FlowProject
+        just involves checking the operation's pre- and post-conditions to
+        determine eligibility. More generally, however, the pre- and
+        post-conditions define a directed acyclic graph that governs the
+        execution of all operations. Visualizing this graph can be useful for
+        finding logic errors in the specified conditions, and having this graph
+        computed also enables additional execution modes. For example, using
+        this graph it is possible to determine exactly what operations need to
+        be executed in order to make the operation eligible so that the task of
+        executing all necessary operations can be automated.
+
+        The graph is determined by iterating over all pairs of operations and
+        checking for equality of pre- and post-conditions. The algorithm builds
+        an adjacency matrix based on whether the pre-conditions for one
+        operation match the post-conditions for another. To ensure robustness,
+        the comparison is relatively strict, requiring that the conditions be
+        the same callable. This means that, for instance, two equivalent (but
+        distinct) lambda functions will not match under this criterion.
+
+        Given a FlowProject subclass defined in a module `project.py`, the
+        output graph could be visualized using Matplotlib and NetworkX with the
+        following code:
+
+        .. code-block:: python
+
+            import numpy as np
+            from matplotlib import pyplot as plt
+            import networkx as nx
+
+            from project import Project
+
+            p = Project()
```

Comment:
I'd prefer to not abbreviate `project` to `p` if this is possible without sacrificing readability.